### PR TITLE
🐛Restrict checking cached ampDoc reference to AMP elements.

### LINKED ISSUE #24995: Form submit service failed: %s: Cannot read property 'fire' of undefined
## What's the issue?

This issue is not happened with the previous version of amp. Amp-form couldn't be submitted because it cannot read property 'fire' of undefined. Further investigated, we noticed an error in amp-form.js amp-form.js:365 Uncaught (in promise) TypeError: this.Ba.whenNextVisible is not a function. The issue was resolved when we rolled back to the previous version of amp, in particular the amp version 1909181902540 doesn't have this issue.
 
## How do we reproduce the issue?

- Open this link:
https://www.greatmats.com/mats/foam-floor-mats.php
- Click on quick calculator
- Enter width and length and submit
- The form cannot be submitted

## What browsers are affected?

All browsers

## Which AMP version is affected?

This is a new issue. We rolled back to the previous 1909181902540 and the form was submitted normally

### PR DESCRIPTION
**Pull request by sparhami**
This fixes a bug where the compiled (renamed) property for `ampDoc_`
could conflict with a form field name. Rather than blacklist `form`,
whitelist `amp-` elements in case anything else has this behavior (like
`window` and Element `id`s).

Fixes #24995

### CODE REVIEW

**Comment by jpettitt** [comment]
Woohoo PR number 25000

**Comment by sparhami** [comment]
> Interesting bug, thanks for investigating and fixing. 
> 
> The way we fixed a similar DOM clobbering issue was to use property names that the validator bans in `form[name]`, `[id]`, etc. I.e.
> 
> ```js
> const cached = node['__amp_doc']; // __amp_* is an invalid value for form[name].
> ```
> 
> We can try extending the window-property-name.js eslint rule to try to catch these generally, though it may have precision problems.

I considered using a string constant, but the problem here is that we are relying on the fact that we are relying on this matching the `ampDoc_` property named here: https://github.com/ampproject/amphtml/blob/b6bd628efa0ad675ffd6fe66526e313640494d37/src/custom-element.js#L171-L175

Since this is set on an `HTMLElement` (since it is a custom element), and we are accessing it from a Node (a super class of HTMLElement), Closure Compiler will ensure consistent property renaming for the set of this property and the read elsewhere.

Perhaps calling it a cached value is not quite correct, since it isn't cached after a lookup. Rather, it is saved when the element is upgraded as we already know which `AmpDoc` it belongs to.

**Comment by sparhami** [comment]
> Another alternative is to avoid accessing private ivars.
> 
> ```js
> const cached = typeof node.getAmpDoc === 'function' ? node.getAmpDoc() : null;
> ```
> 
> Would need a change to dodge the `devAssert`.

I'm not sure how much of a problem this is, since `getAmpDoc` has a `devAssert` requiring the node be connected, if specified. This means that it should also have an `ampDoc_`. The one case where this doesn't hold is the `getAmpDocIfAvailable` entry point (which only has one usage), which explicitly does not require the node to be connected,

**Comment by sparhami** [comment]
Changed to use `getAmpDoc`, but also guarded with `everAttached`, so we should never hit the `devAssert` where `ampDoc_` is `null`.

File: src/service/ampdoc-impl.js [hunk 1]
```diff
@@ -31,6 +31,7 @@ import {isExperimentOn} from '../experiments';
 import {map} from '../utils/object';
 import {parseQueryString} from '../url';
 import {rootNodeFor, waitForBodyOpenPromise} from '../dom';
+import {startsWith} from '../string';
```
Comment (choumx) [inline src/service/ampdoc-impl.js]: No longer needed.

**Pull request merged**